File: scripts/scrape-materials.py
#!/usr/bin/env python3
"""
Scrape SL Hardwoods products from WooCommerce API
Focuses on sheet materials and edging products
"""

import requests
import json
import xml.etree.ElementTree as ET
from urllib.parse import urlparse, parse_qs
import time
from pathlib import Path
import re

# Configuration
BASE_URL = "https://www.slhardwoods.co.uk"
API_BASE = f"{BASE_URL}/wp-json/wc/store/v1/products"
SITEMAP_URLS = [
    f"{BASE_URL}/product-sitemap.xml",
    f"{BASE_URL}/product-sitemap2.xml"
]

# Categories we're interested in
RELEVANT_CATEGORIES = [
    'sheet-material',
    'mdf',
    'plywood',
    'birch-and-hardwood-plywood',
    'melamine',
    'veneered',
    'edging',
    'edge-banding',
    'iron-on-edging'
]

def fetch_sitemap_urls(sitemap_url):
    """Fetch all product URLs from sitemap"""
    print(f"Fetching sitemap: {sitemap_url}")
    try:
        response = requests.get(sitemap_url, timeout=30)
        response.raise_for_status()

        root = ET.fromstring(response.content)
        # Handle namespace
        namespace = {'ns': 'http://www.sitemaps.org/schemas/sitemap/0.9'}
        urls = [loc.text for loc in root.findall('.//ns:loc', namespace)]

        print(f"Found {len(urls)} URLs in sitemap")
        return urls
    except Exception as e:
        print(f"Error fetching sitemap: {e}")
        return []

def is_relevant_product(product_url):
    """Check if product URL is for materials we want"""
    # Skip non-product URLs
    if '/product/' not in product_url:
        return False

    # Check if URL contains relevant categories
    for category in RELEVANT_CATEGORIES:
        if category in product_url.lower():
            return True

    # Also include products with these keywords
    keywords = ['mdf', 'plywood', 'ply', 'veneer', 'melamine', 'birch',
                'oak', 'ash', 'walnut', 'maple', 'cherry', 'edging', 'edge']
    product_slug = product_url.split('/product/')[-1].replace('/', '')

    for keyword in keywords:
        if keyword in product_slug.lower():
            return True

    return False

def extract_product_id_from_url(product_url):
    """Try to extract product ID from URL or fetch the page to get it"""
    # The API needs the product ID, which we can get from the HTML page
    try:
        response = requests.get(product_url, timeout=30)
        response.raise_for_status()

        # Look for product ID in the page source
        # WooCommerce typically includes it in data attributes or scripts
        content = response.text

        # Try to find post ID
        match = re.search(r'post-(\d+)', content)
        if match:
            return match.group(1)

        # Try to find in data-product_id
        match = re.search(r'data-product_id="(\d+)"', content)
        if match:
            return match.group(1)

        return None
    except Exception as e:
        print(f"Error extracting product ID from {product_url}: {e}")
        return None

def fetch_product_data(product_id):
    """Fetch product data from WooCommerce API"""
    try:
        url = f"{API_BASE}/{product_id}"
        response = requests.get(url, timeout=30)
        response.raise_for_status()
        return response.json()
    except Exception as e:
        print(f"Error fetching product {product_id}: {e}")
        return None

def process_product(product_data):
    """Extract relevant information from product data"""
    if not product_data:
        return None

    # Extract basic info
    product_info = {
        'id': product_data.get('id'),
        'name': product_data.get('name', ''),
        'description': product_data.get('description', ''),
        'short_description': product_data.get('short_description', ''),
        'sku': product_data.get('sku', ''),
        'type': product_data.get('type', ''),
        'categories': [],
        'images': [],
        'price': None,
        'price_range': {},
        'variations': [],
        'attributes': []
    }

    # Extract categories
    if 'categories' in product_data:
        product_info['categories'] = [cat.get('name', '') for cat in product_data.get('categories', [])]

    # Extract images
    if 'images' in product_data:
        for img in product_data.get('images', []):
            product_info['images'].append({
                'src': img.get('src', ''),
                'name': img.get('name', ''),
                'alt': img.get('alt', '')
            })

    # Extract pricing
    if 'prices' in product_data:
        prices = product_data['prices']
        product_info['price'] = {
            'regular': prices.get('regular_price', ''),
            'sale': prices.get('price', ''),
            'currency': prices.get('currency_code', 'GBP')
        }

        if 'price_range' in prices:
            product_info['price_range'] = prices['price_range']

    # Extract attributes (like thickness)
    if 'attributes' in product_data:
        for attr in product_data.get('attributes', []):
            product_info['attributes'].append({
                'name': attr.get('name', ''),
                'terms': attr.get('terms', [])
            })

    # Extract variation IDs
    if 'variations' in product_data:
        product_info['variation_ids'] = product_data['variations']

    return product_info

def fetch_variation_data(variation_id):
    """Fetch variation-specific data"""
    try:
        url = f"{API_BASE}/{variation_id}"
        response = requests.get(url, timeout=30)
        response.raise_for_status()
        data = response.json()

        variation_info = {
            'id': data.get('id'),
            'name': data.get('name', ''),
            'sku': data.get('sku', ''),
            'attributes': [],
            'price': None
        }

        # Extract attributes (thickness, etc.)
        if 'attributes' in data:
            for attr in data.get('attributes', []):
                variation_info['attributes'].append({
                    'name': attr.get('name', ''),
                    'value': attr.get('value', '')
                })

        # Extract pricing
        if 'prices' in data:
            prices = data['prices']
            variation_info['price'] = {
                'regular': prices.get('regular_price', ''),
                'sale': prices.get('price', ''),
                'currency': prices.get('currency_code', 'GBP')
            }

        return variation_info
    except Exception as e:
        print(f"Error fetching variation {variation_id}: {e}")
        return None

def main():
    """Main scraping function"""
    print("Starting SL Hardwoods material scraper...")

    # Collect all product URLs
    all_urls = []
    for sitemap_url in SITEMAP_URLS:
        urls = fetch_sitemap_urls(sitemap_url)
        all_urls.extend(urls)

    print(f"Total URLs found: {len(all_urls)}")

    # Filter for relevant products
    relevant_urls = [url for url in all_urls if is_relevant_product(url)]
    print(f"Relevant product URLs: {len(relevant_urls)}")

    # Process products
    all_products = []

    for i, product_url in enumerate(relevant_urls[:50], 1):  # Limit to 50 for testing
        print(f"\n[{i}/{min(50, len(relevant_urls))}] Processing: {product_url}")

        # Extract product ID
        product_id = extract_product_id_from_url(product_url)
        if not product_id:
            print(f"  Could not extract product ID")
            continue

        # Fetch product data
        product_data = fetch_product_data(product_id)
        if not product_data:
            continue

        # Process product
        product_info = process_product(product_data)
        if not product_info:
            continue

        print(f"  Product: {product_info['name']}")
        print(f"  Type: {product_info['type']}")

        # If variable product, fetch all variations
        if product_info['type'] == 'variation' and 'variation_ids' in product_info:
            print(f"  Fetching {len(product_info['variation_ids'])} variations...")
            variations = []

            for var_id in product_info['variation_ids']:
                var_data = fetch_variation_data(var_id)
                if var_data:
                    variations.append(var_data)
                time.sleep(0.5)  # Rate limiting

            product_info['variations'] = variations
            print(f"  Loaded {len(variations)} variations")

        all_products.append(product_info)
        time.sleep(1)  # Rate limiting

    # Save to JSON
    output_dir = Path(__file__).parent.parent / 'public' / 'data'
    output_dir.mkdir(parents=True, exist_ok=True)
    output_file = output_dir / 'scraped-materials.json'

    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(all_products, f, indent=2, ensure_ascii=False)

    print(f"\n✅ Scraped {len(all_products)} products")
    print(f"💾 Saved to: {output_file}")

    # Print summary
    print("\n📊 Summary:")
    print(f"  Total products: {len(all_products)}")
    variable_count = sum(1 for p in all_products if p['type'] == 'variation')
    print(f"  Variable products: {variable_count}")
    simple_count = len(all_products) - variable_count
    print(f"  Simple products: {simple_count}")

if __name__ == '__main__':
    main()

File: scripts/scrape-materials.mjs
#!/usr/bin/env node

/**
 * Scrape SL Hardwoods products from WooCommerce Store API
 * Focuses on sheet materials and edging products
 */

import fetch from 'node-fetch';
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Configuration
const BASE_URL = 'https://www.slhardwoods.co.uk';
const API_BASE = `${BASE_URL}/wp-json/wc/store/v1`;

// Categories we're interested in (based on your requirements)
const MATERIAL_KEYWORDS = [
  'mdf', 'plywood', 'ply', 'veneer', 'melamine', 'birch',
  'oak', 'ash', 'walnut', 'maple', 'cherry', 'beech',
  'moisture resistant', 'mr mdf', 'wbp', 'edge', 'edging'
];

const EDGING_KEYWORDS = [
  'edge', 'edging', 'iron on', 'iron-on', 'lipping'
];

// Helper to add delay
const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));

// Helper to check if product is relevant
function isRelevantProduct(product) {
  const nameLC = product.name.toLowerCase();
  const descLC = (product.description || '').toLowerCase();
  const shortDescLC = (product.short_description || '').toLowerCase();

  // Check if it's a material or edging product
  for (const keyword of MATERIAL_KEYWORDS) {
    if (nameLC.includes(keyword) || descLC.includes(keyword) || shortDescLC.includes(keyword)) {
      return true;
    }
  }

  // Check categories
  if (product.categories && product.categories.length > 0) {
    for (const cat of product.categories) {
      const catName = cat.name.toLowerCase();
      if (catName.includes('sheet') || catName.includes('material') ||
          catName.includes('mdf') || catName.includes('plywood') ||
          catName.includes('edging') || catName.includes('veneered')) {
        return true;
      }
    }
  }

  return false;
}

// Determine material type
function getMaterialType(product) {
  const nameLC = product.name.toLowerCase();

  // Check for edging
  for (const keyword of EDGING_KEYWORDS) {
    if (nameLC.includes(keyword)) {
      return 'EDGING';
    }
  }

  // Check for specific material types
  if (nameLC.includes('birch') && nameLC.includes('ply')) {
    return 'CARCASS'; // Birch plywood is good for carcasses
  }

  if (nameLC.includes('moisture resistant') || nameLC.includes('mr mdf')) {
    return 'CARCASS'; // MR MDF is good for carcasses
  }

  if (nameLC.includes('veneered')) {
    return 'LINING'; // Veneered panels are typically for visible surfaces
  }

  if (nameLC.includes('drawer')) {
    return 'DRAWER';
  }

  // Default to general material
  return 'MATERIAL';
}

// Fetch all products with pagination
async function fetchAllProducts() {
  console.log('🔍 Fetching products from WooCommerce Store API...\n');

  let allProducts = [];
  let page = 1;
  let hasMore = true;
  const perPage = 100; // API default max

  while (hasMore) {
    try {
      const url = `${API_BASE}/products?per_page=${perPage}&page=${page}`;
      console.log(`📄 Fetching page ${page}...`);
      console.log(`   URL: ${url}`);

      const response = await fetch(url, {
        headers: {
          'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
          'Accept': 'application/json',
        }
      });

      console.log(`   Status: ${response.status} ${response.statusText}`);
      console.log(`   Content-Type: ${response.headers.get('content-type')}`);

      if (!response.ok) {
        if (response.status === 400 || response.status === 404) {
          // No more pages
          console.log('✅ Reached end of products\n');
          hasMore = false;
          break;
        }
        const text = await response.text();
        console.log(`   Response body preview: ${text.substring(0, 500)}`);
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      const contentType = response.headers.get('content-type');
      if (!contentType || !contentType.includes('application/json')) {
        const text = await response.text();
        console.log(`   Not JSON response. Content-Type: ${contentType}`);
        console.log(`   Response preview: ${text.substring(0, 500)}`);
        hasMore = false;
        break;
      }

      const products = await response.json();

      if (!products || products.length === 0) {
        console.log('✅ No more products found\n');
        hasMore = false;
        break;
      }

      console.log(`   Found ${products.length} products`);
      allProducts.push(...products);

      page++;
      await delay(1000); // Rate limiting - 1 second between requests

    } catch (error) {
      console.error(`❌ Error fetching page ${page}:`, error.message);
      hasMore = false;
    }
  }

  console.log(`\n📊 Total products fetched: ${allProducts.length}\n`);
  return allProducts;
}

// Process product data
function processProduct(product) {
  const processed = {
    id: product.id,
    name: product.name,
    slug: product.slug,
    type: product.type, // simple, variable, variation
    sku: product.sku || '',
    description: product.description || '',
    short_description: product.short_description || '',
    material_type: getMaterialType(product),
    categories: [],
    images: [],
    price: null,
    variations: []
  };

  // Extract categories
  if (product.categories && Array.isArray(product.categories)) {
    processed.categories = product.categories.map(cat => cat.name);
  }

  // Extract images
  if (product.images && Array.isArray(product.images)) {
    processed.images = product.images.map(img => ({
      src: img.src,
      name: img.name || '',
      alt: img.alt || product.name
    }));
  }

  // Extract pricing
  if (product.prices) {
    const prices = product.prices;
    processed.price = {
      regular: prices.regular_price || '',
      sale: prices.price || '',
      currency: prices.currency_code || 'GBP',
      formatted: prices.price_range ? {
        min: prices.price_range.min_amount || '',
        max: prices.price_range.max_amount || ''
      } : null
    };
  }

  // Extract attributes (e.g., thickness)
  if (product.attributes && Array.isArray(product.attributes)) {
    processed.attributes = product.attributes.map(attr => ({
      name: attr.name,
      terms: attr.terms || []
    }));
  }

  return processed;
}

// Main scraping function
async function main() {
  console.log('🌲 SL Hardwoods Material Scraper (Node.js)\n');
  console.log('=' .repeat(60) + '\n');

  try {
    // Fetch all products
    const allProducts = await fetchAllProducts();

    // Filter for relevant products (sheet materials and edging)
    console.log('🔍 Filtering for sheet materials and edging...\n');
    const relevantProducts = allProducts.filter(isRelevantProduct);
    console.log(`✅ Found ${relevantProducts.length} relevant products\n`);

    // Process products
    console.log('⚙️  Processing product data...\n');
    const processedProducts = relevantProducts.map((product, index) => {
      if ((index + 1) % 50 === 0) {
        console.log(`   Processed ${index + 1}/${relevantProducts.length} products...`);
      }
      return processProduct(product);
    });

    // Categorize by material type
    const categorized = {
      carcass: processedProducts.filter(p => p.material_type === 'CARCASS'),
      lining: processedProducts.filter(p => p.material_type === 'LINING'),
      drawer: processedProducts.filter(p => p.material_type === 'DRAWER'),
      edging: processedProducts.filter(p => p.material_type === 'EDGING'),
      general: processedProducts.filter(p => p.material_type === 'MATERIAL')
    };

    console.log('\n📊 Material Type Breakdown:');
    console.log(`   Carcass materials: ${categorized.carcass.length}`);
    console.log(`   Lining materials: ${categorized.lining.length}`);
    console.log(`   Drawer materials: ${categorized.drawer.length}`);
    console.log(`   Edging products: ${categorized.edging.length}`);
    console.log(`   General materials: ${categorized.general.length}`);

    // Save to JSON
    const outputDir = path.join(__dirname, '..', 'public', 'data');
    await fs.mkdir(outputDir, { recursive: true });

    const outputFile = path.join(outputDir, 'scraped-materials.json');
    await fs.writeFile(
      outputFile,
      JSON.stringify(processedProducts, null, 2),
      'utf8'
    );

    const categorizedFile = path.join(outputDir, 'materials-by-type.json');
    await fs.writeFile(
      categorizedFile,
      JSON.stringify(categorized, null, 2),
      'utf8'
    );

    console.log(`\n✅ Scraped ${processedProducts.length} materials`);
    console.log(`💾 Saved to:`);
    console.log(`   - ${outputFile}`);
    console.log(`   - ${categorizedFile}`);

    // Show some examples
    if (processedProducts.length > 0) {
      console.log(`\n📝 Sample products:\n`);
      processedProducts.slice(0, 5).forEach((p, i) => {
        console.log(`${i + 1}. ${p.name}`);
        console.log(`   Type: ${p.material_type}`);
        console.log(`   Price: ${p.price?.formatted ?
          `£${p.price.formatted.min} - £${p.price.formatted.max}` :
          p.price?.sale || 'N/A'}`);
        console.log(`   Images: ${p.images.length}`);
        console.log('');
      });
    }

  } catch (error) {
    console.error('\n❌ Fatal error:', error);
    process.exit(1);
  }
}

// Run the scraper
main().catch(console.error);
